lib/spack/spack/build_environment.py

```diff
@@ -185,40 +190,62 @@ def set_build_environment_variables(pkg, env, dirty):
         dirty (bool): Skip unsetting the user's environment settings
     """
     # Gather information about various types of dependencies
-    build_deps      = set(pkg.spec.dependencies(deptype=('build', 'test')))
-    link_deps       = set(pkg.spec.traverse(root=False, deptype=('link')))
-    build_link_deps = build_deps | link_deps
+    build_deps      = pkg.spec.dependencies(deptype=('build', 'test'))
+    link_deps       = list(pkg.spec.traverse(root=False, deptype=('link')))
+    build_link_deps = dedupe(build_deps + link_deps)
     rpath_deps      = get_rpath_deps(pkg)
 
-    build_prefixes      = [dep.prefix for dep in build_deps]
-    link_prefixes       = [dep.prefix for dep in link_deps]
-    build_link_prefixes = [dep.prefix for dep in build_link_deps]
-    rpath_prefixes      = [dep.prefix for dep in rpath_deps]
-
     # add run-time dependencies of direct build-time dependencies:
-    for build_dep in build_deps:
+    for build_dep in list(build_deps):
         for run_dep in build_dep.traverse(deptype='run'):
-            build_prefixes.append(run_dep.prefix)
+            build_deps.append(run_dep)
+    build_deps = dedupe(build_deps)
 
     # Filter out system paths: ['/', '/usr', '/usr/local']
     # These paths can be introduced into the build when an external package
     # is added as a dependency. The problem with these paths is that they often
     # contain hundreds of other packages installed in the same directory.
     # If these paths come first, they can overshadow Spack installations.
-    build_prefixes      = filter_system_paths(build_prefixes)
-    link_prefixes       = filter_system_paths(link_prefixes)
-    build_link_prefixes = filter_system_paths(build_link_prefixes)
-    rpath_prefixes      = filter_system_paths(rpath_prefixes)
-
-    # Prefixes of all of the package's dependencies go in SPACK_DEPENDENCIES
-    env.set_path(SPACK_DEPENDENCIES, build_link_prefixes)
+    def filter_system_deps(deps):
```

Comment:
that's exactly this: 

https://github.com/LLNL/spack/blob/develop/lib/spack/spack/util/environment.py#L46-L47

why substitute?